File: tools/pythonpkg/tests/fast/test_extensions.py
import duckdb
import os
import glob
import pandas as pd
from os.path import abspath

class TestExtensions(object):
    def test_extensions(self, duckdb_cursor):

        # Paths to search for extensions
        extension_search_patterns = [
            "../../../../build/release/extension/*/*.duckdb_extension",
            "../../*.duckdb_extension"
        ]

        # DUCKDB_PYTHON_TEST_EXTENSION_PATH can be used to add a path for the extension test to search for extensions
        if 'DUCKDB_PYTHON_TEST_EXTENSION_PATH' in os.environ:
            env_extension_path = os.getenv('DUCKDB_PYTHON_TEST_EXTENSION_PATH');
            env_extension_path = env_extension_path.rstrip('/')
            extension_search_patterns.append(env_extension_path + '/*/*.duckdb_extension')
            extension_search_patterns.append(env_extension_path + '/*.duckdb_extension')

        # Depending on the env var, the test will fail on not finding any extensions
        must_test_extension_load = os.getenv('DUCKDB_PYTHON_TEST_EXTENSION_REQUIRED', False)

        extension_paths_found = []
        for pattern in extension_search_patterns:
            extension_pattern_abs = abspath(pattern)
            print(f"Searching path: {extension_pattern_abs}")
            for path in glob.glob(extension_pattern_abs):
                extension_paths_found.append(path)

        def test_httpfs(duckdb_cursor):
            try:
                duckdb_cursor.execute("SELECT id, first_name, last_name FROM PARQUET_SCAN('https://raw.githubusercontent.com/cwida/duckdb/master/data/parquet-testing/userdata1.parquet') LIMIT 3;")
            except RuntimeError as e:
                # Test will ignore result if it fails due to networking issues while running the test.
                if (str(e).startswith("HTTP HEAD error")):
                    return
                elif (str(e).startswith("Unable to connect")):
                    return
                else:
                    raise e

            result_df = duckdb_cursor.fetchdf()
            exp_result = pd.DataFrame({
                'id': pd.Series([1, 2, 3], dtype="int32"),
                'first_name': ['Amanda', 'Albert', 'Evelyn'],
                'last_name': ['Jordan', 'Freeman', 'Morgan']
            })
            assert(result_df.equals(exp_result))

        if (must_test_extension_load and not extension_paths_found):
            raise Exception("Env var DUCKDB_PYTHON_TEST_EXTENSION_REQUIRED was set, but no extensions were found to test with!");

        for path in extension_paths_found:
            conn = duckdb.connect()
            conn.execute(f"LOAD '{path}'")

            if (path.endswith("httpfs.duckdb_extension")):
                test_httpfs(conn)

File: tools/rpkg/tests/testthat/test_extensions.R
#skip_on_cran()

library("DBI")
library("testthat")

load_extension <- function(extension_name) {
  con <- dbConnect(duckdb::duckdb(config=list("allow_unsigned_extensions"="true")))
  matches <- list.files(pattern='*.duckdb_extension', recursive=FALSE, full.names=TRUE)
  matches <- c(matches, list.files(path='/tmp/duckdb_extensions', pattern='*.duckdb_extension', recursive=TRUE, full.names=TRUE))
  matches <- c(matches, list.files(path=system.file(package = 'duckdb'), pattern='*.duckdb_extension', recursive=TRUE, full.names=TRUE))

  require_variable <- Sys.getenv("DUCKDB_R_TEST_EXTENSION_REQUIRED")

  if (require_variable == "1") {
    expect_true(length(matches) > 0, "DUCKDB_R_TEST_EXTENSION_REQUIRED was set, but no extensions were found")
  } else {
    if (length(matches) == 0) {
        warning("No extensions found for testing!")
    }
  }
  found <- FALSE
  for (extension in matches){
    # Test loading the extension
    # HTTPFS extension specific test
    if (grepl(extension_name, extension, fixed = TRUE)) {
        query <- paste("LOAD '", extension, "';", sep = "")
        dbExecute(con, query)
        found <-TRUE
    }
  }
  if (found != TRUE){
  warning(paste("Required extension ", extension_name, "not found"))
  }
  return (con)
}

test_that("extensions can be loaded", {
    skip_if(Sys.getenv("DUCKDB_R_TEST_EXTENSION_REQUIRED") != "1", "DUCKDB_R_TEST_EXTENSION_REQUIRED not set, hence not testing extensions")
    con <- load_extension("httpfs.duckdb_extension")
    on.exit(dbDisconnect(con, shutdown = TRUE))
    df0 <- data.frame(
         id = c(1, 2, 3),
         first_name = c("Amanda", "Albert", "Evelyn"),
         last_name = c("Jordan", "Freeman", "Morgan"),
         stringsAsFactors = FALSE
       )
    res <- dbGetQuery(con, "SELECT id, first_name, last_name FROM PARQUET_SCAN('https://raw.githubusercontent.com/cwida/duckdb/master/data/parquet-testing/userdata1.parquet') LIMIT 3;")
    expect_equal(df0, res)
})

test_that("substrait extension test", {
  skip_if(Sys.getenv("DUCKDB_R_TEST_EXTENSION_REQUIRED") != "1", "DUCKDB_R_TEST_EXTENSION_REQUIRED not set, hence not testing extensions")
  con <- load_extension("substrait.duckdb_extension")
  on.exit(dbDisconnect(con, shutdown = TRUE))
  dbExecute(con, "CREATE TABLE integers (i INTEGER)")
  dbExecute(con, "INSERT INTO integers VALUES (42)")
  plan <- duckdb::duckdb_get_substrait(con, "select * from integers limit 5")
  result <- duckdb::duckdb_prepare_substrait(con, plan)
  df <- dbFetch(result)
  expect_equal(df$i, 42L)

  result_arrow <- duckdb::duckdb_prepare_substrait(con, plan, TRUE)
  df2 <- as.data.frame(duckdb::duckdb_fetch_arrow(result_arrow))
  expect_equal(df2$i, 42L)
})

